sls-versions/src/main/java/com/palantir/sls/versions/HandRolledMatcherParser.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.sls.versions;
+
+import java.util.Optional;
+import java.util.OptionalInt;
+
+/**
+ * A hand-rolled implementation of {@link RegexMatcherParser}.
+ *
+ * Uses two-element int[] arrays as the 'result' type to represent a successful/failed parse, where result[0] always
+ * contains the index into the string we're looking at, and result[1] contains the value returned by the parser
+ * combinator.
+ */
+final class HandRolledMatcherParser {
```

Comment:
I'm open to the argument that this is overkill... 

I kinda just wanted to write it to see how good we could make this, given that this class is so heavily used inside Apollo. The use of jqwik is really what tips this over into being viable imo - I've found that changing _any_ line in this file will trip tests... I'm pretty confident this cannot diverge from the reference impl

Reply:
Haven't reviewed the logic yet, but I'd rename this to `SlsVersionMatcherParser`. The previous implementation was hand-rolled to delegate to the java regex implementation with a specific pattern ;-)